File: Regular Expression Matching.java
public class Solution {
    public boolean isMatch(String s, String p) {
        if (s.isEmpty() && p.isEmpty())
            return true;
        if (p.isEmpty())
            return false;
        if (s.isEmpty())
            return (p.length() >= 2 && p.charAt(1) == '*' && isMatch(s, p.substring(2)) );
        
        if (p.length() >= 2 && p.charAt(1) == '*') 
            return isMatch(s, p.substring(2)) || (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.') && isMatch(s.substring(1), p);
        if (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.')
            return isMatch(s.substring(1), p.substring(1));
        return false;
    }
}

File: Regular Expression Matching.cpp
class Solution {
public:
    bool isMatch(const char *s, const char *p) {
        if (*s == 0 && *p == 0)
            return true;
        if (*p == 0)
            return false;
        if (*(p+1) == '*') 
            return ((*s == *p || *p == '.') && (*s != 0) && isMatch(s+1, p)) || isMatch(s, p+2); 
        if (*s != 0 && (*s == *p || *p == '.'))
            return isMatch(s+1, p+1);
        return false;
    }
};